lib/pcap/source.ex

```diff
@@ -1,54 +1,136 @@
+
+  @next_packet &Parser.next_packet/1
 
   def_output_pads output: [
                     caps: :any
                   ]
 
   def_options packet_transformer: [
                 type: :function,
-                spec: (any() -> Buffer.t()),
-                default: &__MODULE__.identity/1,
+                spec: (Packet.t() -> Buffer.t()),
+                default: &__MODULE__.default_transformer/1,
                 description: """
-                This function must transform parsed packet into a Buffer.
+                This function transforms parsed packet into a buffer.
+                It is applied on each packet.
+
+                It allows enriching buffers with metadata extracted from
+                packets f.e. packet source address or port.
                 """
               ],
-              file: [
+              path: [
                 type: :string,
                 description: "Path to the .pcap file"
               ]
 
   defmodule State do
-    @enforce_keys [:data, :transformer]
-    defstruct @enforce_keys
+    @moduledoc false
+    @enforce_keys [:transformer, :path]
+    defstruct @enforce_keys ++ [:parser]
+
+    @type t :: %__MODULE__{
+            transformer: (Packet.t() -> Buffer.t()),
+            path: binary()
+          }
   end
 
   @impl true
-  def handle_init(%__MODULE__{file: file_path, packet_transformer: transformer}) do
-    data = ExPcap.from_file(file_path).packets
-    {:ok, %State{data: data, transformer: transformer}}
+  def handle_init(%__MODULE__{path: path, packet_transformer: transformer}) do
+    %State{
+      path: path,
+      transformer: transformer
+    }
+    ~> {:ok, &1}
   end
 
   @impl true
-  def handle_demand(pad, size, unit, context, state)
+  def handle_prepared_to_playing(_context, %State{path: path} = state) do
+    case Parser.from_file(path) do
+      {:ok, parser} -> {:ok, %State{state | parser: parser}}
+      {:error, _} = error -> {error, state}
+    end
+  end
 
-  def handle_demand(:output, _size, :buffers, _ctx, %State{data: []} = state) do
-    {{:ok, [{:event, {:output, %EndOfStream{}}}]}, state}
+  @impl true
+  def handle_prepared_to_stopped(_context, %State{parser: parser} = state) do
+    Parser.destroy(parser)
+    {:ok, %State{state | parser: nil}}
   end
 
+  @impl true
   def handle_demand(:output, size, :buffers, _ctx, state) do
-    %State{data: data, transformer: transformer} = state
-    {out, rest} = Enum.split(data, size)
-    buffers = Enum.map(out, fn packet -> transformer.(packet) end)
-    action = [buffer: {:output, buffers}]
-    {{:ok, action}, %State{state | data: rest}}
+    %State{parser: parser, transformer: transformer} = state
+
+    size
+    |> fetch_packets(parser, transformer)
+    |> handle_fetched_packets(state)
   end
 
-  # TODO change me in case of rewrite to streaming
-  @spec identity(any()) :: any()
-  def identity(%ExPcap.Packet{parsed_packet_data: {_, payload}}),
+  @spec default_transformer(any()) :: any()
```

Comment:
still not fixed

Reply:
If you are talking about the name I replied in the discussion previously.

Reply:
i'm talking about the types

Reply:
Mea maxima culpa I misunderstood.